adafruit_usb_descriptor/hid.py

```diff
@@ -0,0 +1,269 @@
+# in the Software without restriction, including without limitation the rights
+# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in
+# all copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+# THE SOFTWARE.
+
+from . import util
+import struct
+
+"""
+HID specific descriptors
+========================
+
+* Author(s): Dan Halbert
+"""
+
+HID_CLASS = 0x03
+
+# Many other subclasses omitted.
+
+HID_SUBCLASS_NOBOOT = 0x00
+HID_SUBCLASS_BOOT = 0x01
+
+HID_PROTOCOL_GENERIC = 0x00
+HID_PROTOCOL_KEYBOARD = 0x01
+HID_PROTOCOL_MOUSE = 0x02
+
+class HIDDescriptor:
+    """Lists upcoming HID report descriptors."""
+    bDescriptorType = 0x21
+    fmt = "<BB" + "HBBBH"
+    bLength = struct.calcsize(fmt)
+
+    def __init__(self, *,
+                 description="unknown HIDDescriptor",
+                 bcdHID=0x1101,
+                 bCountryCode=0x0,
+                 bNumDescriptors=1,
+                 bDescriptorType_Class=0x22,
+                 wDescriptorLength):
+        self.description = description
+        self.bcdHID = bcdHID
+        self.bCountryCode = bCountryCode
+        self.bNumDescriptors = bNumDescriptors
+        self.bDescriptorType_Class = bDescriptorType_Class
+        self.wDescriptorLength = wDescriptorLength
+
+    def __bytes__(self):
+        return struct.pack(self.fmt,
+                           self.bLength,
+                           self.bDescriptorType,
+                           self.bcdHID,
+                           self.bCountryCode,
+                           self.bNumDescriptors,
+                           self.bDescriptorType_Class,
+                           self.wDescriptorLength)
+
+class ReportDescriptor:
+    """Describes multiple kinds of reports sent by this HID device.
+    """
+
+    def __init__(self, *,
+                 description="unknown ReportDescriptor",
+                 report_descriptor):
+        self.description = description
+        self.report_descriptor = report_descriptor
+
+    def __bytes__(self):
+        return self.report_descriptor
+
+ReportDescriptor.GENERIC_MOUSE_REPORT = ReportDescriptor(
+    description="GENERIC_MOUSE_REPORT",
+    report_descriptor=bytes([
+        0x05, 0x01,     # Usage Page (Generic Desktop),
```

Comment:
It'd be more memory efficient if these were byte strings rather than lists that get converted. I believe it'd be:

```python
report_descriptor=
b'\x05\x01' # Usage Page (Generic Desktop),
b'\x09\x02' # Usage (Mouse),
...
```

You may need parens around it too.

Reply:
The comma-separated format is as-generated by several different tools I have and also online examples. So it's easier to edit if I leave it in that format. Actually running this takes a fraction of a second.

Reply:
I'm thinking ahead to when we do use it in CircuitPython. Anyway, this is good for now. We can tweak it later if need be.